File: pi-extension/src/skills.ts
/**
 * skills.ts — JSONL-backed skill (macro) library.
 *
 * Mirror of the mainline skills.py with the same on-disk format so a
 * skill recorded by either side can be replayed by either side.  Each
 * line in `skills.jsonl` is one record:
 *
 *   {
 *     "name": str,
 *     "keywords": [str, ...],
 *     "app": str,
 *     "steps": [{"tool": str, "args": object}, ...],
 *     "created": float,        // unix seconds
 *     "success_count": int
 *   }
 *
 * Append-only writes; full rewrite is done atomically (tmp file + rename)
 * for `delete` and `incrementSuccess`.
 */

import { mkdir, readFile, rename, writeFile } from "node:fs/promises";
import { dirname } from "node:path";

export interface SkillStep {
  tool: string;
  args: Record<string, unknown>;
}

export interface Skill {
  name: string;
  keywords: string[];
  app: string;
  steps: SkillStep[];
  created: number;
  success_count: number;
}

function tokenize(s: string): string[] {
  return (s || "")
    .toLowerCase()
    .split(/\W+/)
    .filter((t) => t.length > 2);
}

export class SkillStore {
  constructor(private readonly path: string) {}

  async all(): Promise<Skill[]> {
    try {
      const text = await readFile(this.path, "utf8");
      const skills: Skill[] = [];
      for (const line of text.split(/\r?\n/)) {
        if (!line.trim()) continue;
        try {
          skills.push(JSON.parse(line) as Skill);
        } catch {
          // skip malformed lines
        }
      }
      return skills;
    } catch (e) {
      if ((e as NodeJS.ErrnoException).code === "ENOENT") return [];
      throw e;
    }
  }

  async get(name: string): Promise<Skill | undefined> {
    return (await this.all()).find((s) => s.name === name);
  }

  async search(query: string, limit = 5): Promise<Skill[]> {
    const all = await this.all();
    if (!query.trim()) return all.slice(0, limit);
    const qtoks = new Set(tokenize(query));
    if (qtoks.size === 0) return [];
    const scored: Array<{ score: number; success: number; created: number; skill: Skill }> = [];
    for (const s of all) {
      const ktoks = new Set([...tokenize(s.keywords.join(" ")), ...tokenize(s.name)]);
      let overlap = 0;
      for (const t of qtoks) if (ktoks.has(t)) overlap++;
      if (overlap === 0 && !s.name.toLowerCase().includes(query.toLowerCase())) continue;
      scored.push({ score: -overlap, success: -(s.success_count ?? 0), created: -(s.created ?? 0), skill: s });
    }
    scored.sort((a, b) =>
      a.score - b.score || a.success - b.success || a.created - b.created,
    );
    return scored.slice(0, limit).map((x) => x.skill);
  }

  async save(input: { name: string; keywords?: string[]; app?: string; steps: SkillStep[] }): Promise<Skill> {
    if (!input.name) throw new Error("Skill name is required");
    if (!input.steps?.length) throw new Error("Cannot save a skill with no steps");
    const existing = (await this.all()).filter((s) => s.name !== input.name);
    const skill: Skill = {
      name: input.name,
      keywords: input.keywords ?? [],
      app: input.app ?? "",
      steps: input.steps,
      created: Date.now() / 1000,
      success_count: 0,
    };
    existing.push(skill);
    await this.rewrite(existing);
    return skill;
  }

  async incrementSuccess(name: string): Promise<void> {
    const all = await this.all();
    let changed = false;
    for (const s of all) {
      if (s.name === name) {
        s.success_count = (s.success_count ?? 0) + 1;
        changed = true;
      }
    }
    if (changed) await this.rewrite(all);
  }

  async delete(name: string): Promise<boolean> {
    const all = await this.all();
    const kept = all.filter((s) => s.name !== name);
    if (kept.length === all.length) return false;
    await this.rewrite(kept);
    return true;
  }

  private async rewrite(skills: Skill[]): Promise<void> {
    await mkdir(dirname(this.path), { recursive: true });
    const tmp = this.path + ".tmp";
    const body = skills.map((s) => JSON.stringify(s)).join("\n") + (skills.length ? "\n" : "");
    await writeFile(tmp, body, "utf8");
    await rename(tmp, this.path);
  }
}

// ---------------------------------------------------------------------------
// Step normalization
// ---------------------------------------------------------------------------

const ACTIVATE_TOOLS = new Set([
  "desktop_launch",
  "desktop_focus_window",
  "desktop_activate_window", // mainline name — cross-compat
]);
const TYPE_TOOLS = new Set(["desktop_type", "desktop_hotkey"]);

/**
 * Remove pixel-coordinate focus clicks sandwiched between a window-activation
 * step and a type step.
 *
 * Saved skills sometimes include a desktop_click(x, y) immediately after a
 * focus/launch step as a focus gesture. On replay the window may open at a
 * different screen position, so the hardcoded coordinates miss the window and
 * steal focus before desktop_type fires. The preceding activate step already
 * established focus, so the click can be dropped safely.
 */
export function normalizeSkillSteps(steps: SkillStep[]): SkillStep[] {
  const drop = new Set<number>();
  for (let i = 0; i < steps.length; i++) {
    const step = steps[i];
    const args = step.args as Record<string, unknown>;
    if (
      step.tool === "desktop_click"
      && typeof args["x"] === "number"
      && typeof args["y"] === "number"
      && i > 0
      && i + 1 < steps.length
      && ACTIVATE_TOOLS.has(steps[i - 1].tool)
      && TYPE_TOOLS.has(steps[i + 1].tool)
    ) {
      drop.add(i);
    }
  }
  return steps.filter((_, i) => !drop.has(i));
}

File: skills.py
"""
skills.py — Persistent skill (macro) library.

A "skill" is a named, ordered list of tool calls — typically captured from a
successful task — that can be replayed later either by the agent (when the
model decides the task matches) or by the standalone replay CLI (no LLM).

Storage is plain JSONL at ~/.autogui/skills.jsonl, one record per line:

    {"name": str,
     "keywords": [str, ...],
     "app": str,
     "steps": [{"tool": str, "args": dict}, ...],
     "created": float,        # unix timestamp
     "success_count": int}

The append-only format keeps the data trivially inspectable, mergeable, and
diffable without locking — good enough for a single-user agent.
"""

from __future__ import annotations

import json
import logging
import re
import time
from pathlib import Path

logger = logging.getLogger(__name__)


def _tokenize(text: str) -> list[str]:
    return [t for t in re.split(r"\W+", (text or "").lower()) if t and len(t) > 2]


# ---------------------------------------------------------------------------
# Step normalization
# ---------------------------------------------------------------------------

_ACTIVATE_TOOLS = frozenset({"desktop_launch", "desktop_activate_window"})
_TYPE_TOOLS = frozenset({"desktop_type", "desktop_hotkey"})


def normalize_skill_steps(steps: list[dict]) -> list[dict]:
    """Remove pixel-coordinate focus clicks sandwiched between a window-activation
    step and a type step.

    Saved skills sometimes include a desktop_click(x, y) immediately after
    desktop_activate_window or desktop_launch as a focus gesture.  On replay
    the window may open at a different screen position, so the hardcoded
    coordinates miss the window and steal focus before desktop_type fires.
    Dropping the click is safe because the preceding activate step already
    established focus.
    """
    drop: set[int] = set()
    for i, step in enumerate(steps):
        tool = step.get("tool", "")
        args = step.get("args") or {}
        if (
            tool == "desktop_click"
            and "x" in args and "y" in args
            and i > 0
            and i + 1 < len(steps)
        ):
            prev_tool = steps[i - 1].get("tool", "")
            next_tool = steps[i + 1].get("tool", "")
            if prev_tool in _ACTIVATE_TOOLS and next_tool in _TYPE_TOOLS:
                drop.add(i)
                logger.debug(
                    "[skills] normalize: dropping pixel focus-click at step %d (x=%s, y=%s)",
                    i, args.get("x"), args.get("y"),
                )
    return [s for i, s in enumerate(steps) if i not in drop]


class SkillStore:
    """JSONL-backed skill library with simple keyword retrieval."""

    def __init__(self, path: str = "~/.autogui/skills.jsonl"):
        self.path = Path(path).expanduser()
        self.path.parent.mkdir(parents=True, exist_ok=True)
        if not self.path.exists():
            self.path.touch()

    # ------------------------------------------------------------------
    # Read
    # ------------------------------------------------------------------

    def all(self) -> list[dict]:
        skills: list[dict] = []
        try:
            for line in self.path.read_text(encoding="utf-8").splitlines():
                if not line.strip():
                    continue
                try:
                    skills.append(json.loads(line))
                except json.JSONDecodeError:
                    logger.warning("[skills] Skipping malformed line in %s", self.path)
        except FileNotFoundError:
            return []
        return skills

    def get(self, name: str) -> dict | None:
        for s in self.all():
            if s.get("name") == name:
                return s
        return None

    def search(self, query: str, limit: int = 5) -> list[dict]:
        """Score by keyword overlap; ties broken by success_count then recency."""
        if not query:
            return self.all()[:limit]
        qtoks = set(_tokenize(query))
        if not qtoks:
            return []
        scored: list[tuple[int, int, float, dict]] = []
        for s in self.all():
            ktoks = set(_tokenize(" ".join(s.get("keywords", [])) + " " + s.get("name", "")))
            overlap = len(qtoks & ktoks)
            if overlap == 0 and query.lower() not in s.get("name", "").lower():
                continue
            scored.append((
                -overlap,
                -int(s.get("success_count", 0)),
                -float(s.get("created", 0.0)),
                s,
            ))
        scored.sort()
        return [s for *_, s in scored[:limit]]

    # ------------------------------------------------------------------
    # Write
    # ------------------------------------------------------------------

    def save(
        self,
        *,
        name: str,
        keywords: list[str],
        app: str,
        steps: list[dict],
    ) -> dict:
        if not name:
            raise ValueError("Skill name is required")
        if not steps:
            raise ValueError("Cannot save a skill with no steps")
        # Replace any existing skill with the same name.
        existing = [s for s in self.all() if s.get("name") != name]
        skill = {
            "name": name,
            "keywords": list(keywords or []),
            "app": app or "",
            "steps": list(steps),
            "created": time.time(),
            "success_count": 0,
        }
        existing.append(skill)
        self._rewrite(existing)
        return skill

    def increment_success(self, name: str):
        skills = self.all()
        changed = False
        for s in skills:
            if s.get("name") == name:
                s["success_count"] = int(s.get("success_count", 0)) + 1
                changed = True
        if changed:
            self._rewrite(skills)

    def delete(self, name: str) -> bool:
        skills = self.all()
        kept = [s for s in skills if s.get("name") != name]
        if len(kept) == len(skills):
            return False
        self._rewrite(kept)
        return True

    def _rewrite(self, skills: list[dict]):
        # Atomic rewrite to avoid corrupting the file mid-write.
        tmp = self.path.with_suffix(self.path.suffix + ".tmp")
        with tmp.open("w", encoding="utf-8") as f:
            for s in skills:
                f.write(json.dumps(s, ensure_ascii=False) + "\n")
        tmp.replace(self.path)
